specification/containerservice/Fleet.Management/tspconfig.yaml

```
@@ -2,7 +2,9 @@ emit:
   - '@azure-tools/typespec-autorest'
 options:
   '@azure-tools/typespec-autorest':
-    azure-resource-provider-folder: '{cwd}/../resource-manager'
+    arm-types-dir: ../../../../../../common-types/resource-management
```

Is this necessary? That path looks like a recipe for trouble.

---

```suggestion
    arm-types-dir: "{project-root}/../../common-types/resource-management"
```
it should be using `{project-root}` (not sure I fixed it correctly here)

---

We should not have a relative path here.  We should be able to use a path starting from "{project-root}" https://github.com/Azure/typespec-azure/blob/main/packages/typespec-autorest/src/openapi.ts#L799-L808

---

I merged @timotheeguerin suggestion, CI pipeline will tell us if it works.

---

Does this mean that folks with this reference will not be able to run outside of the specs repo? For the most part we have blocked references outside of the service directory but this will be one which goes outside of it.

---

@timotheeguerin, @markcowl: The change from
```
arm-types-dir: ../../../../../../common-types/resource-management
```
to:
```
arm-types-dir: "{project-root}/../../common-types/resource-management"
```

Is causing a diff in the generated swagger (extra `..`):

```
-            "$ref": "../../../../../../common-types/resource-management/v3/types.json#/parameters/ApiVersionParameter"
+            "$ref": "../../../../../../../common-types/resource-management/v3/types.json#/parameters/ApiVersionParameter"
```

https://dev.azure.com/azure-sdk/public/_build/results?buildId=2788701&view=logs&j=9cd39ef8-47c1-5b58-08db-8c3708712f1b&t=462145f3-d740-50de-51e9-126e7c0a29e2&l=43

Is the new path correct or incorrect?

---

Sorry I was not sure about my exact fix. I was just saying for reliability it is better to use the absolute path relative to the project root y than counting on where the openapi json file gets generated

---

@weshaggard would mean that if you are using arm outside of this repo you do need to give a reference of where those types which could be a url or a copy of it

---

What is the location of `{project-root}`?

---

Found this in the docs:

```
project-root - Points to the the tspconfig.yaml file containing folder.
```

---

I wonder if this is a bug in TypeSpec.  If the docs are correct, then `{project-root}` should point to:

```
specification/containerservice/Fleet.Management
```

The common types are at:

```
specification/common-types/resource-management/(v1|v2|v3|v4|v5)
```

So this should be correct (up 2 dirs from `{project-root}`):

```
arm-types-dir: "{project-root}/../../common-types/resource-management"
```

The swagger is generated in this folder:

```
specification/containerservice/resource-manager/Microsoft.ContainerService/fleet/preview/2023-03-15-preview/fleets.json
```

So the common types are 6 dirs up from the generated swagger folder.  Which suggests the old swagger was correct (6 `..`), and the new swagger is wrong (7 `..`):

```
-            "$ref": "../../../../../../common-types/resource-management/v3/types.json#/parameters/ApiVersionParameter"
+            "$ref": "../../../../../../../common-types/resource-management/v3/types.json#/parameters/ApiVersionParameter"
```

---

Hhm totally possible, could it be related to the other config issue I fixed recently?

---

Are you just using tsp compile . ?

---

Yes, you can see the exact commands run by the CI pipeline here:

https://dev.azure.com/azure-sdk/public/_build/results?buildId=2788701&view=logs&j=9cd39ef8-47c1-5b58-08db-8c3708712f1b&t=462145f3-d740-50de-51e9-126e7c0a29e2&l=15

---

This change works, so this does look like a bug in either the docs or the product for `{project-root}`.  It appears that 
`{project-root}` is actually pointing to the folder one above the folder containing `tspconfig.yaml`.

```
-    arm-types-dir: "{project-root}/../../common-types/resource-management"
+    arm-types-dir: "{project-root}/../common-types/resource-management"
```

---

yeah guessing bug in autorest emitter. `project-root` should be the directory where tspconfig.yaml is.

and I guess you can verify that because the `emitter-output-dir` is currently `project-root/..`, is that generating correctly?

---

Yes, `emitter-output-dir` works correctly.  Can you open an issue for this against the autorest emitter?

What do you recommend we use for now to unblock this PR?  Revert back to using `../../../../../..` instead of `{project-root}`?

---

yeah I would say even though relative is confusing it is at least correct. Fixing this bug will create a breaking change if using `project-root`

---

Yeah, I don't want to use `{project-root}/..` since it will break when the bug is fixed.

---

created issue https://github.com/Azure/typespec-azure/issues/3024

---

migth be worthy of a hotfix

---

hhm weird I can't repro by making a new project with the same structure

---

OK, I actually found this issue when looking at a customer issue, glad you guys found it here.  It looks like we are off by one directory in interpolation.  I think this is worthy of a hotfix, because we really want to use the project directory - this allows us to have one boilerplate setting for arm-types-dir, and let the actual output directory dictate the relative path from spec to common files.

---

nvm, repro'd I copied the wrong tspconfig

---

fix here https://github.com/Azure/typespec-azure/pull/3025

---

We can remove this config then as it is defualt value.

---

I will create a PR to upgrade the specs repo to use the new autorest emitter (and verify it doesn't regress any specs currently building)

---

PR to update tools: https://github.com/Azure/azure-rest-api-specs/pull/24100

Pipeline `specs - typespec - ci` is passing, which means no spec in the repo was currently relying on the buggy behavior.  So we can either:

1. Merge this PR, then 24100, then fix the usage of `{project-root}` in a third PR
2. Merge 24100, then fix the usage of `{project-root}` in this PR, then merge this PR